File: 0143-reorder-list/0143-reorder-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *   int val;
 *   ListNode next;
 *   ListNode() {}
 *   ListNode(int val) { this.val = val; }
 *   ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
  public void reorderList(ListNode head) {
    ListNode slow = head;
    ListNode fast = head.next;
    while (fast != null && fast.next != null) {
      slow = slow.next;
      fast = fast.next.next;
    }
    ListNode second = slow.next;
    slow.next = null;
    ListNode prev = null;
    ListNode next;
    while (second != null) {
      next = second.next;
      second.next = prev;
      prev = second;
      second = next;
    }
    ListNode first = head;
    second = prev;
    ListNode next1;
    ListNode next2;
    while (second != null) {
      next1 = first.next;
      next2 = second.next;
      first.next = second;
      second.next = next1;
      first = next1;
      second = next2;
    }
  }
}

File: 0143-reorder-list/0143-reorder-list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *   Val int
 *   Next *ListNode
 * }
 */
func reorderList(head *ListNode)  {
  slow, fast := head, (*head).Next
  for fast != nil && (*fast).Next != nil {
    slow, fast = (*slow).Next, (*(*fast).Next).Next
  }
  second := (*slow).Next
  (*slow).Next = nil
  var prev, next *ListNode
  for second != nil {
    next = (*second).Next
    (*second).Next = prev
    prev = second
    second = next
  }
  first := head
  second = prev
  var next1, next2 *ListNode
  for second != nil {
    next1, next2 = (*first).Next, (*second).Next
    (*first).Next, (*second).Next = second, next1
    first = next1
    second = next2
  }
}